### PR TITLE
Resolve for specific platforms

### LINKED ISSUE #4295: bundle package flag --all-platforms has no effect
When i run `bundle package --all-platforms` I would expect `vendor/cache` to be populated with `.gem` for all platforms, however I only get the current platform gem.

```
# bundle env
Environment

    Bundler   1.11.2
    Rubygems  2.5.1
    Ruby      2.3.0p0 (2015-12-25 revision 53290) [x86_64-linux]
    GEM_HOME  /usr/local/bundle
    GEM_PATH
    Git       2.1.4

Bundler settings

    path
      Set for the current user (/root/.bundle/config): "/usr/local/bundle"
    bin
      Set for the current user (/root/.bundle/config): "/usr/local/bundle/bin"
    silence_root_warning
      Set for the current user (/root/.bundle/config): "true"
    app_config
      Set via BUNDLE_APP_CONFIG: "/usr/local/bundle"

Gemfile

    source 'https://rubygems.org'

    gem 'google-protobuf', '3.0.0.alpha.5.0.3'

Gemfile.lock

    GEM
      remote: https://rubygems.org/
      specs:
        google-protobuf (3.0.0.alpha.5.0.3)

    PLATFORMS
      ruby

    DEPENDENCIES
      google-protobuf (= 3.0.0.alpha.5.0.3)

    BUNDLED WITH
       1.11.2
```

and

```
# bundle package --all-platforms --verbose
HTTP GET https://bundler.rubygems.org/api/v1/dependencies
HTTP 200 OK
Fetching gem metadata from https://rubygems.org/
Query List: ["google-protobuf"]
Query Gemcutter Dependency Endpoint API: google-protobuf
HTTP GET https://bundler.rubygems.org/api/v1/dependencies?gems=google-protobuf
HTTP 200 OK
Query List: []
Fetching version metadata from https://rubygems.org/
Query List: []
Installing google-protobuf 3.0.0.alpha.5.0.3
0:  google-protobuf (3.0.0.alpha.5.0.3) from /usr/local/bundle/specifications/google-protobuf-3.0.0.alpha.5.0.3-x86_64-linux.gemspec
Using bundler 1.11.2
0:  bundler (1.11.2) from /usr/local/bundle/specifications/bundler-1.11.2.gemspec
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Bundled gems are installed into /usr/local/bundle.
Updating files in vendor/cache
  * google-protobuf-3.0.0.alpha.5.0.3-x86_64-linux.gem
```

My ultimate desire would be to just get the `ruby` platform version and build from source, but I thought that `--all-platforms` would be the next best alternative. I couldn't find this reported elsewhere (but was mentioned in https://github.com/bundler/bundler-features/issues/4#issuecomment-76508244)

### PR DESCRIPTION
**Pull request by segiddins**
Closes #4295.

This will require adding a bunch of tests, as well as figuring out how to put this new behavior behind a feature flag (thus fixing all of the existing tests).

### CODE REVIEW

File: lib/bundler/dependency.rb [hunk 1]
```diff
@@ -52,7 +52,7 @@ class Dependency < Gem::Dependency
       :x64_mingw_20 => Gem::Platform::X64_MINGW,
       :x64_mingw_21 => Gem::Platform::X64_MINGW,
       :x64_mingw_22 => Gem::Platform::X64_MINGW,
-      :x64_mingw_23 => Gem::Platform::X64_MINGW
+      :x64_mingw_23 => Gem::Platform::X64_MINGW,
```
Comment (indirect) [inline lib/bundler/dependency.rb]: really? we're doing trailing commas now? 😝

File: lib/bundler/rubygems_ext.rb [hunk 2]
```diff
@@ -167,6 +167,9 @@ class Platform
     MSWIN64 = Gem::Platform.new("mswin64") unless defined?(MSWIN64)
     MINGW = Gem::Platform.new("x86-mingw32") unless defined?(MINGW)
     X64_MINGW = Gem::Platform.new("x64-mingw32") unless defined?(X64_MINGW)
+    # DARWIN = Gem::Platform.new("universal-darwin") unless defined?(DARWIN)
+    # LINUX = Gem::Platform.new("x86-linux") unless defined?(LINUX)
+    # X64_LINUX = Gem::Platform.new("x86_64-linux") unless defined?(X64_LINUX)
```
Comment (indirect) [inline lib/bundler/rubygems_ext.rb]: commented out but included deliberately?

File: lib/bundler/source.rb [hunk 3]
```diff
@@ -14,6 +14,7 @@ def unmet_deps
 
     def version_message(spec)
       message = "#{spec.name} #{spec.version}"
+      message += " (#{spec.platform})" if Bundler.ui.debug? && spec.platform != Gem::Platform::RUBY
```
Comment (indirect) [inline lib/bundler/source.rb]: we probably want this always, even when debug is off

File: lib/bundler/spec_set.rb [hunk 4]
```diff
@@ -99,6 +99,17 @@ def materialize(deps, missing_specs = nil)
       SpecSet.new(materialized.compact)
     end
 
+    def materialized_for_all_platforms
```
Comment (indirect) [inline lib/bundler/spec_set.rb]: is the all platforms here every platform currently listed in the gemfile?

File: lib/bundler/gem_helpers.rb [hunk 5]
```diff
@@ -28,5 +28,38 @@ def generic_local_platform
...
+          1
+        end,
+
+        if spec_platform.version == user_platform.version
+          0
+        elsif spec_platform.version.nil?
+          1
+        else
+          2
+        end,
```
Comment (indirect) [inline lib/bundler/gem_helpers.rb]: please turn each of these into comparison methods or something 😎

**Comment by indirect** [comment]
omg this is the most hopeful I have felt about platforms in literally five years, no exaggeration. hooray :tada:

File: lib/bundler/dependency.rb (hunk 1 above)
Comment (segiddins) [inline lib/bundler/dependency.rb]: we have been since introducing rubocop

File: lib/bundler/rubygems_ext.rb (hunk 2 above)
Comment (segiddins) [inline lib/bundler/rubygems_ext.rb]: nope, part of the process to get to this point, it can be deleted

File: lib/bundler/source.rb (hunk 3 above)
Comment (segiddins) [inline lib/bundler/source.rb]: 👍 I was just trying to avoid changing the output

File: lib/bundler/spec_set.rb (hunk 4 above)
Comment (segiddins) [inline lib/bundler/spec_set.rb]: it's just all the specs in the spec set, regardless of what platform they're for, in contrast to how `for` does platform filtering (and specifically different from how `materialize` calls `for` only for the current platform)

File: lib/bundler/dependency.rb (hunk 1 above)
Comment (indirect) [inline lib/bundler/dependency.rb]: :sob:

Comment (segiddins) [inline lib/bundler/dependency.rb]: this way we don't have to touch that line when adding a new element to the array

Comment (indirect) [inline lib/bundler/dependency.rb]: what good is it to be in charge of a project if we adopt code style that clashes with my personal coding aesthetic 😝

**Comment by indirect** [comment]
@homu r+

**Comment by indirect** [comment]
@homu r-

**Comment by indirect** [comment]
@homu r+

**Comment by indirect** [comment]
@homu r-

**Comment by indirect** [comment]
(Waiting to merge this until after we stable-branch 1.13)

**Comment by mwrock** [comment]
This fixes #4896

File: lib/bundler/definition.rb [hunk 6]
```diff
@@ -105,8 +107,9 @@ def initialize(lockfile, dependencies, sources, unlock, ruby_version = nil, opti
 
       @gem_version_promoter = create_gem_version_promoter
 
-      current_platform = Bundler.rubygems.platforms.map {|p| generic(p) }.compact.last
-      add_platform(current_platform)
+      current_platform = Bundler.rubygems.platforms.last
+      add_platform(current_platform) if Bundler.settings[:specific_platform]
```
Comment (segiddins) [inline lib/bundler/definition.rb]: we'll probably need to special-case around when in deployment mode and not add any platforms and count on the platforms check in https://github.com/bundler/bundler/pull/4895

**Comment by indirect** [comment]
@homu r+

**Comment by segiddins** [comment]
@homu r- until I address my last comment

**Comment by segiddins** [comment]
@indirect this should finally be mergable

**Comment by indirect** [comment]
hooray 🏅 @homu r+

**Comment by segiddins** [comment]
Rebase'd
@homu r=indirect

**Pull request merged**